org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMDocTypeFormatter.java

```diff
@@ -175,4 +195,43 @@ private void removeLeftSpaces(int to, List<TextEdit> edits) {
 		formatterDocument.removeLeftSpaces(to, edits);
 	}
 
+	private String getQuotationAsString() {
+		return formatterDocument.getSharedSettings().getPreferences().getQuotationAsString();
+	}
+
+	private EnforceQuoteStyle getEnforceQuoteStyle() {
+		return formatterDocument.getSharedSettings().getFormattingSettings().getEnforceQuoteStyle();
+	}
+
+	private int getIdStart(DOMDocumentType docType) {
+		if (docType.getPublicIdNode() != null) {
+			return docType.getPublicIdNode().getStart();
+		} else if (docType.getSystemIdNode() != null) {
+			return docType.getSystemIdNode().getStart();
+		} else
+			return 0;
+	}
+
+	private int getIdEnd(DOMDocumentType docType) {
+		if (docType.getPublicIdNode() != null) {
+			return docType.getPublicIdNode().getEnd();
+		} else if (docType.getSystemIdNode() != null) {
+			return docType.getSystemIdNode().getEnd();
+		} else
+			return 0;
+	}
+
+	private void replaceQuoteWithPreferred(DTDDeclNode nodeDecl, DTDDeclParameter parameter, List<TextEdit> edits){
+		int paramStart = parameter.getStart();
+		int paramEnd = parameter.getEnd();
+		if (StringUtils.isQuoted(nodeDecl.getOwnerDocument().getText().substring(paramStart, paramEnd))) {

```

Comment:
You can do this check by looping chatacters to avoid using substring which creates a new String instance which is not good for memory performance.